src/hooks/useDebounce.js

```diff
@@ -2,6 +2,10 @@ import {useEffect, useRef} from 'react';
 import lodashDebounce from 'lodash/debounce';
 
 /**
+ * Create and return a debounced function.
+ * 
+ * Make sure to pass a stable function reference to prevent recreating the debounced function on each render.
```

Comment:
> recreating the debounced function on each render.

I don't think that that's the point. The API user might think  that `func` is the "same" function from the debouncing algorithm perspective, even if it actually _changes_, let's say from _f1_ to _f2_. Someone might expected. I the denounced function is called `d`, and `wait` = 100 ms, we can sketch this scenario:

- `func` = `f1` (initial)
- `d(...args1)` (timer starts, waiting for 100ms...)
- 20 ms pass...
- `d(...args2)` (timer resets, waiting for 100ms...)
- 100 ms pass... 
- `f1(...args2)`
- some time passes... 
- `d(...args3)` (timer starts, waiting for 100ms...)
- 50 ms pass...
- `func` = `f2`
- **Possibly expected behavior:**
  - 50 ms pass....
  - `f2(...args3)`
- **Actual behavior:**
  - _nothing_

Reply:
Maybe...

> Every time the identity of any of the arguments changes, the debounce operation will restart (canceling any ongoing debounce). This is especially important in the case of `func`. To prevent that, pass stable references.

Writing this comment made me realize we have a bigger problem here, i.e. the `options` argument...

Reply:
Jeez, I hate JavaScript and React. Don't tell anyone, though.

Reply:
Solutions:
- either just remove it from `useDebounce` and add a warning comment
- or unpack the `options` to the set known options at this time, depend on all of them, and re-pack them in the invocation

Reply:
There are just 3 options, so maybe let's repack?

Reply:
https://stackoverflow.com/questions/54095994/react-useeffect-comparing-objects

Reply:
Unless something changed, React is using `===` to compare dependency values. It's definitely not deep equality, it's not even shallow equality, this just plain equality.

```js
({leading: true}) !== ({leading: true})
```

Reply:
So we can repack those properties:

```js
const {leading, maxWait, trailing} = options;

//  Use leading, maxWait, trailing in deps and the call
```

Reply:
> Sorry, I still don't understand the issue.

The issue is that passing an object to React deps is a mistake. It's ineffective, as two objects (constructed through object literals) are never equal to each other. In consequence, **every** render in the parent function component using `useDebounce` would reset the timer.

> If yes, then useMemo can solve it, right?

It would be an abuse of `useMemo`, as it's officially only for computation-heavy tasks and there's no guarantee that a future version of React won't drop & recompute `useMemo`'s value at an arbitrary time. See [the docs](https://react.dev/reference/react/useMemo#caveats).

Reply:
Constructing an object through object literal would be considered pure for many purposes, but _not_ when we depend on its identity ("address"), like in the case when it's passed to `deps`.

Reply:
@bernhardoj

In that case, the object returned by the first `useMemo`...

```js
  const searchOptions = useMemo(() => {
    return { matchMode: 'whole-word', text };
  }, [text]); // ✅ Only changes when text changes
```

...is passed to _another_ `useMemo`:

```js
  const visibleItems = useMemo(() => {
    return searchItems(allItems, searchOptions);
  }, [allItems, searchOptions]); // ✅ Only changes when allItems or searchOptions changes
```

As we already said, the function inside `useMemo` is expected to be pure and there are no strict guarantees about how many times it will be called. In consequence, there are also no strict requirements about the stability of the dependencies, as far as we talk about _correctness_. It starts to matter only when we start to discuss _performance_.

This **changes** when we pass such an object to a dependency of `useEffect`, which _does_ care about how and when the dependencies change, because, well, an _effect_ is executed every time it happens.

If that's what was confusing, maybe I can redact my previous statement and make it...

_Constructing an object through object literal would be considered pure for many purposes, but not when we depend on its identity ("address"), like in the case when it's passed to `useEffect` deps._

Reply:
Having said all that, I agree that this is a messy topic, and I wouldn't be shocked even if React docs themselves, or some official examples, give contradicting information.